packages/form-js-playground/src/components/PlaygroundRoot.js

```diff
@@ -120,11 +123,24 @@ export function PlaygroundRoot(props) {
       emit('formPlayground.rendered');
     });
 
-    form.on('changed', event => {
-      setResultData(event.data);
-    });
+    /**
+     * Set submit form data to result view.
+     *
+     * This makes sure we do not end up in an
+     * endless loop as submit validation also
+     * changes the form.
+     */
+    const _updateResultData = () => {
+      form.off('changed', _updateResultData);
```

Comment:
I would go against this method personally, toggling events for exceptions feels like an unstable workaround that will cause problems in the future.

Isn't the problem here that we're calling the validation in the first place? This wasn't ever part of the update cycle, and I still don't think it should. In general, submitting the form is a lot of side effects that we don't want to execute every render. 

So I don't think we care about validation at all, so can we instead extract the part of the submit function responsible for building the data property that's being deconstructed L135 ([ie this aggregator here](https://github.com/bpmn-io/form-js/blob/7a876f52cb8fdc8f7143d0c1c61d7587d6f8cc20/packages/form-js-viewer/src/Form.js#L172)) into its own function and pass that around instead? Call it `buildSubmissionData`, and then you don't need to deconstruct, and you don't need the workaround.

Reply:
Thanks for the feedback! I thought about this as well and I guess my preference was to ensure we do not break the existing submit behavior, to ensure we get the correct data state during that time.

What you describe makes a lot of sense as [`form.validate`](https://github.com/bpmn-io/form-js/blob/develop/packages/form-js-viewer/src/Form.js#L216)  does not change the form data, just the error state. Therefore I could also imagine to simply wrap the bare submitting functionality without the validation part and use that one instead.

I will have another look on Thursday :+1:

Reply:
I extracted the relevant code into an own method via https://github.com/bpmn-io/form-js/pull/364/commits/a18062cc3f565d9f295708f19011ffe2bffecdd2. The rationale why I use an internal method of the form, instead of public or simply a util:

I think that's internal behavior and consumers should instead use `submit` to get the full submit experience. We can use this internally in other packages to keep things consistent, though (e.g. form-js-playground).